bcdaworker/queueing/manager/alr.go

```diff
@@ -21,52 +31,139 @@ type alrQueue struct {
 	ctx := context.Background()
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	// Unmarchall JSON that contains the job details
-	// JSON selected for ALR for continuity... may change in the future
+	// Unmarshall JSON that contains the job details
 	var jobArgs models.JobAlrEnqueueArgs
 	err := json.Unmarshal(job.Args, &jobArgs)
+	// If we cannot unmarshall this, it would be very problematic.
 	if err != nil {
-		q.alrLog.Warnf("Failed to unmarhall job.Args '%s' %s. Removing job...",
+		// TODO: perhaps just fail the job?
+		q.alrLog.Warnf("Failed to unmarhall job.Args '%s' %s.",
 			job.Args, err)
-		// By returning nil to que-go, job is removed from queue
-		return nil
 	}
 
 	// Check if the job was cancelled
-	go func() {
-		for {
-			select {
-			case <-time.After(15 * time.Second):
-				jobStatus, err := q.repository.GetJobByID(ctx, jobArgs.ID)
-
-				if err != nil {
-					q.alrLog.Warnf("Could not find job %d status: %s", jobArgs.ID, err)
-				}
-
-				if jobStatus.Status == models.JobStatusCancelled {
-					// cancelled context will get picked up by worker.go#writeBBDataToFile
-					cancel()
-					return
-				}
-			case <-ctx.Done():
-				return
+	go checkIfCancelled(ctx, q, cancel, jobArgs)
+
+	// Validate the job like bcdaworker/worker/worker.go#L43
+	// TODO: Abstract this into either a function or interface like the bfd worker
+	alrJobs, err := q.repository.GetJobByID(ctx, jobArgs.ID)
+	if err != nil { // if this is not nil, we could not find the Job
+		// Drill down to what kind of error this is...
+		if errors.Is(err, repository.ErrJobNotFound) {
+			// Parent job is not found
+			// If parent job is not found reach maxretry, fail the job
+			if job.ErrorCount >= maxRetry {
+				q.alrLog.Errorf("No job found for ID: %d acoID: %s. Retries exhausted. Removing job from queue.", jobArgs.ID,
+					jobArgs.CMSID)
+				// By returning a nil error response, we're singaling to que-go to remove this job from the jobqueue.
+				return nil
+			}
+			q.alrLog.Warnf("No Job with ID %d ACO %s found. Will retry.", jobArgs.ID, jobArgs.CMSID)
+			return fmt.Errorf("could not retrieve job from database")
+		}
+		// Else that job just doesn't exist
+		return fmt.Errorf("failed to valiate job: %w", err)
+	}
+	// If the job was cancalled...
+	if alrJobs.Status == models.JobStatusCancelled {
+		q.alrLog.Warnf("ALR big job has been cancelled, worker will not be tasked for %s",
+			job.Args)
+		return nil
+	}
+	// If the job has been failed by a previous worker...
+	if alrJobs.Status == models.JobStatusFailed {
+		q.alrLog.Warnf("ALR big job has been failed, worker will not be tasked for %s",
+			job.Args)
+		return nil
+	}
+	// End of validation
+
+	// Keep track of how many times a small job has failed.
+	// If the threshold is reached, fail the parent job
+	// This is additional logic that does not exist for BFD worker
+	if _, exists := alrJobTracker[alrJobs.ID]; !exists {
```

Comment:
Right now the job tracker is used to see if we need to update the job status to in progress. What if the job is marked as in progress, fails and gets picked up by another work instance? We would update that status to InProgress as well, right?

Does it  make sense to always do an update and use a conditional update like we do in the BFD worker? [It does a check then set](https://github.com/CMSgov/bcda-app/blob/041a24f75c650ffe7dd07597c6b9723c22180bd3/bcdaworker/repository/postgres/repository.go#L118).